_raw should walk the tree

lapointexavier opened the pull request:
Bugfix for recursively calling the `_raw` property on a schema

lapointexavier committed:
repro

lapointexavier committed:
Call _raw recursively to walk the tree

shermany commented:
LGTM

lapointexavier committed:
Actually test both use case

benauthor commented:
@shermany thx for your input

benauthor commented on bfh/__init__.py:
```diff
@@ -130,7 +130,10 @@ def _raw(self):
         for name in self._field_names:
             value = getattr(self, name)
-            out[name] = value
+            if isinstance(value, Schema):
```
We should also cover the situation where a Schema is nested within a GenericSchema. So let's use hasattr here or refactor `_raw` onto SchemaInterface and check on that, and make sure GenericSchema._raw does similar. The end goal is a tree of any mix of Schema and GenericSchema should be transformed into a tree of all GenericSchema.

benauthor replied:
Also we need to account for iterables full of Schema/GenericSchema.

lapointexavier committed:
wip: nested Schema and GenericSchema should behave as expected with All

lapointexavier committed:
Adds test_all_transformation_generic_schema

lapointexavier committed:
Merge branch 'master' into all_strict_bugfix

lapointexavier committed:
Add tests with nested ArrayField

lapointexavier committed:
Minor formatting annoyance

lapointexavier committed:
Refactor of _raw

lapointexavier committed:
Keep it simple

lapointexavier commented:
@benauthor Should be good for another look

benauthor commented on bfh/__init__.py:
```diff
@@ -209,7 +228,12 @@ def is_empty(self):
     def _raw(self):
-        return self
+        out = dict(**self.__dict__)
+        for name in self._field_names:
```
GenericSchema's _field_names will always be empty. This is because it has no fields declared on it -- it's generic.

I'd expect some test to be failing right now. Your tests are pretty elaborate. Perhaps we also need some very basic tests of `_raw` on Schema and GenericSchema?

```
In [2]: wow = bfh.GenericSchema(foo="wow")

In [3]: wow.serialize()
Out[3]: {'foo': 'wow'}

In [4]: wow._field_names
Out[4]: []
```

lapointexavier replied:
Ahhh, right. Missed that one...

benauthor commented on tests/test_transformations.py:
```diff
@@ -127,6 +160,218 @@ class AllStrict(Mapping):
+            user_name = Get('user', 'name')
+            object = All(strict=False)
+
+        class SomeStrictMapping(Mapping):
```
Can you simplify the test of All by omitting the mappings entirely (i.e. look at the other tests in this class)? Can you rename the test to express exactly what you are testing? This feels integration-y and too wide a test.

benauthor commented on tests/test_transformations.py:
```diff
@@ -127,6 +160,218 @@ class AllStrict(Mapping):
+            user_name = Get('user', 'name')
+            object = All(strict=True)
+
+        non_strict = SomeNonStrictMapping().apply(self.basic_data).serialize()
```
I understand you're trying to be DRY but putting the fixtures elsewhere like this feels like a loss for clarity. I can't understand this test at a glance.

benauthor commented on tests/test_transformations.py:
```diff
@@ -127,6 +160,218 @@ class AllStrict(Mapping):
+            name = UnicodeField()
+
+            # All() should know how to deal with a Schema within
+            # a GenericSchema
```
Ditto above; if this is all you're trying to test, then much of this 65 line test is just noise.

benauthor commented on tests/test_transformations.py:
```diff
@@ -127,6 +160,218 @@ class AllStrict(Mapping):
+        self.assertDictEqual({'object': nested_data}, res)
+
+        res = StrictNestedMappingWithIter().apply(nested_data).serialize()
+        self.assertDictEqual({'object': strict_nested_data}, res)
```
ditto

lapointexavier committed:
PR feedback
- Remove noisy tests, will write something simpler
- Fix the _raw property on GenericSchema, also need a test to cover this

lapointexavier committed:
Simple nested schema test

lapointexavier committed:
Unused imports

lapointexavier committed:
Adds tests to show make sure All(strict=False) works on nested GenericSchema / Schema

lapointexavier committed:
Also asserts that what Schema/GenericSchema _raw returns matches the expected data once serialized

lapointexavier committed:
Change for more specific tests directly on _raw

lapointexavier committed:
Stick an array in there to test _get_raw_value properly

lapointexavier committed:
Forgot, don't subclass GenericSchema.

lapointexavier committed:
Test GenericSchema directly as instance, not subclassed

benauthor commented:
👍
